common/src/device/transport/mqtt/ClientWebSocketChannel.cs

```diff
@@ -140,7 +146,7 @@ protected override async void DoBeginRead()
                 do
                 {
                     byteBuffer = allocHandle.Allocate(allocator);
-                    allocHandle.LastBytesRead = await DoReadBytes(byteBuffer).ConfigureAwait(false);
```

Comment:
We were seeing hangs in mqtt runners and the recommendation from Max was that we could never use ConfigureAwait with dot netty. After removing this, we never saw hangs again. We should avoid adding ConfigureAwaits in paths that use their own thread pool

---
In reply to: [217763495](https://github.com/Azure/azure-iot-sdk-csharp/pull/619#discussion_r217763495) [](ancestors = 217763495)

Reply:
>  could never use ConfigureAwait with dot netty. 

@alextolp `ConfigureAwait` is mandatory by FXCop rules. Not setting it will break the build. You can either set it to `true` (return on this context) or `false` (any context).

@nayato I don't understand the reasoning of not using any thread to continue the execution. Up to the first `await` statement the void callback will execute on the original thread. After that, the thread returns to dotNetty. Why spend the cycles to ensure that we're returning on that random thread at a future time?

Reply:
@CIPop the whole point is that it is not a random thread. DotNetty overrides current task scheduler so if you don't opt-out with ConfigureAwait continuation will be posted to executor thread. Primary benefit of doing that is to avoid synchronization on data access and random issues associated with data races when we fail to do so.

Reply:
Thanks @nayato - didn't know that! (Really cool!) 
In that case the right code here would be `.ConfigureAwait(true)`.

Reply:
Do we want to have explicit .ConfigureAwait(true) as it is the default?

---
In reply to: [218609670](https://github.com/Azure/azure-iot-sdk-csharp/pull/619#discussion_r218609670) [](ancestors = 218609670)

Reply:
Yes, otherwise build will break due to the FXCop rule. Our code should be explicit instead of implicit. DotNetty is a special case as it has its own TaskScheduler. Most library code should always do `.ConfigureAwait(false)` to allow completions on the current thread and avoid a thread switch.